File: LeetCode/Rotting Oranges/Rotting Oranges.py
Intutive Approach :
		1. Create a list of all rotten oranges with time elapsed as 0
		2. For all rotten oranges , find their neighboring oranges and make all fresh as rotten 
		3. Increment the timestamp of all the neighbors by +1 simultaneously 
		4. If any fresh oranges left, return -1 
		
		def orangesRotting(grid):
		      
				m = len(grid)
				n = len(grid[0])
				rotten = []
				fresh = 0

                #Fill in values in rotten with all rotten oranges with timestamp as 0 for every entry
				for i in range(0,m):
					for j in range(0,n):
						if grid[i][j] == 2:
							rotten.append((i,j,0))
						elif grid[i][j] ==1:
							fresh+=1
							
                #Edge case where every entry in grid is either 0 or 2
				if fresh ==0:
					return(0)
                
				#Function to return all the neighbors of the current cell
				def neighbors(x,y):
				    #Four directions as stated in question
					dir_x= [-1,1,0,0] 
					dir_y = [0,0,-1,1]

					for i in range(0,4):
						new_x = x+dir_x[i]
						new_y = y+dir_y[i]

						if not (0 <= new_x < m and 0 <= new_y < n): 
							continue

						yield((new_x,new_y))

				seen = set() # To keep track of which cells have been seen as of now 
				
				#Start rotting via BFS
				
				while rotten :
					elem = rotten.pop(0)
					row = elem[0]
					col = elem[1]
					time = elem[2]

					for neighbor in neighbors(row,col):  # Find all the neighbors of current cell
						r = neighbor[0]
						c = neighbor[1]
						if neighbor in seen:   #If already rotten, skip
							continue
						seen.add((r,c))
						if grid[r][c]==1:   #If neighbor is 1 i.e. fresh , then make it rotten i.e.=2
							grid[r][c]=2
							rotten.append((r,c,time+1))  #Increment time simultaneously for all the neighbors of current cell
                 
				#If there is still any fresh oranges left, return -1
				for i in range(0,m):        
					for j in range(0,n):
						if grid[i][j] == 1:
							return(-1)
							
                #return final time elapsed 
				return(time)
				

File: LeetCode/Rotting Oranges/Rotting Oranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int n = grid.size();
        int m = grid[0].size();
        int ans=0,freshCount=0; 
        
        int dir[5] = {-1,0,1,0,-1}; 
        
        // using a dir array which is useful to travel in four directions i.e top left bottom right
        
            
        queue<pair<int,pair<int,int>>> q;

        // using queue to apply bfs on grid used type of queue as pair<int,pair<int,int>>
        // so in queue three things are stored first the mintime and then row and column index of the grid cell

        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j] == 2){ // initially push all the rotten oranges in queue and mark it visited i.e grid[i][j] = -1
                    grid[i][j] = -1;
                    q.push({0,{i,j}});
                }
                if(grid[i][j] == 1)freshCount++; // keeping track of count of fresh oranges in freshCount variable
            }
        }

        // here logic applied is to apply bfs and in each one unit time for each rotten orange 
        // make adjacent neighbouring fresh oranges rotten and push the rotten orange cell
        // row and column index and mintime+1 in queue and repeat the process until all adjacent
        // fresh oranges are rotten 
        
        while(!q.empty()){
            int qsize = q.size();
            
            for(int i=0;i<qsize;i++){
                ans = q.front().first;
                int row = q.front().second.first;
                int col = q.front().second.second;
                q.pop();
                
                for(int i=0;i<4;i++){
                    int newrow = row + dir[i];
                    int newcol = col + dir[i+1];
                    
                    if(newrow < 0 || newcol < 0 || newrow == n || newcol == m)continue; // condition to check if adjacent cell of grid is within the boundary of grid

                    if(grid[newrow][newcol] == 1){   // if fresh orange found make it rotten mark it as visited and push it in queue
                        freshCount--;
                        grid[newrow][newcol] = -1;
                        q.push({mintime+1,{newrow,newcol}});  
                    }
                }
            }
            
        }

        // return ans only if all the fresh oranges are rotten otherwise return -1

        return (freshCount == 0)?ans:-1; 
    }
};

